subiquity/controllers/filesystem.py

```diff
@@ -129,16 +129,24 @@ def __init__(self, app):
         self.answers.setdefault('guided-index', 0)
         self.answers.setdefault('manual', [])
         self._cur_probe = None
+        self._monitor = None
         self.ui_shown = False
 
     def start(self):
-        urwid.connect_signal(
-            self.app, 'debug-shell-exited', self._maybe_reprobe_block)
         self._start_probe(restricted=False)
-
-    def _maybe_reprobe_block(self):
-        if not self.ui_shown:
+        context = pyudev.Context()
+        self._monitor = pyudev.Monitor.from_netlink(context)
+        self._monitor.filter_by(subsystem='block')
+        self._monitor.enable_receiving()
+        self._last_udev_event = time.time()
+        self.loop.watch_file(self._monitor.fileno(), self._udev_event)
+
+    def _udev_event(self):
+        action, dev = self._monitor.receive_device()
+        log.debug("_udev_event %s %s", action, dev)
+        if time.time() - self._last_udev_event > 0.1 and not self.ui_shown:
```

Comment:
Since we get the event, and we know it's block related, why not use:

udevadm settle --exit-if-exists=dev

This blocks until udev has finished processing the events related to dev?

Reply:
My thinking was that udev events tend to come in a flurry and if we just wait for a period of time with no events that's good enough. Waiting for the udevadm queue to settle is probably better though (it's what the network monitoring code does). Whatever we do though, this logic is busted, it will only process the first event of a flurry. Let me fix that...